src/ddx_driver.f90

```diff
@@ -9,200 +9,353 @@
+        do_g = .false.
+    end if
 end if
 
-! allocate for everything
-! it is possible to pass to mkrhs unallocated arrays as optional arguments
-! but only since fortran 2008
-allocate(phi_cav(ddx_data % constants % ncav), &
-    & gradphi_cav(3, ddx_data % constants % ncav), &
-    & hessianphi_cav(3, 3, ddx_data % constants % ncav), &
-    & force(3, ddx_data % params % nsph), &
-    & psi(ddx_data % constants % nbasis, ddx_data % params % nsph))
+! Allocate the arrays for the required electrostatic properties.
+if (do_phi) then
+    allocate(phi_cav(ddx_data % constants % ncav), stat=info)
+    if (info .ne. 0) then
+        write(6, *) "Allocation failed in ddx_driver"
+        stop 1
+    end if
+end if
+if (do_e) then
+    allocate(e_cav(3, ddx_data % constants % ncav), stat=info)
+    if (info .ne. 0) then
+        write(6, *) "Allocation failed in ddx_driver"
+        stop 1
+    end if
+end if
+if (do_g) then
+    allocate(g_cav(3, 3, ddx_data % constants % ncav), stat=info)
+    if (info .ne. 0) then
+        write(6, *) "Allocation failed in ddx_driver"
+        stop 1
+    end if
+end if
+
+! Compute the required electrostatic properties.
+if (do_phi .and. do_e .and. do_g) then
+    call build_g(ddx_data % params, ddx_data % constants, &
+        & ddx_data % workspace, multipoles, 0, phi_cav, e_cav, g_cav)
+else if (do_phi .and. do_e) then
+    call build_e(ddx_data % params, ddx_data % constants, &
+        & ddx_data % workspace, multipoles, 0, phi_cav, e_cav)
+else
+    call build_phi(ddx_data % params, ddx_data % constants, &
+        & ddx_data % workspace, multipoles, 0, phi_cav)
+end if
 
+finish_time = omp_get_wtime()
+write(*, 100) "Electrostatic properties time:", finish_time-start_time, &
+    & " seconds"
+
+! STEP 3b: adjoint RHS.
 start_time = omp_get_wtime()
-call mkrhs(ddx_data % params, ddx_data % constants, ddx_data % workspace, &
-    & phi_flag, phi_cav, grad_flag, gradphi_cav, hessian_flag, &
-    & hessianphi_cav, psi, charges)
+allocate(psi(ddx_data % constants % nbasis, ddx_data % params % nsph), &
+    & stat=info)
+if (info .ne. 0) then
+    write(6, *) "Allocation failed in ddx_driver"
+    stop 1
+end if
+call build_psi(ddx_data % params, multipoles, 0, psi)
 finish_time = omp_get_wtime()
-write(*, "(A,ES11.4E2,A)") " mkrhs time:", finish_time-start_time, " seconds"
+write(*, 100) "Psi time:", finish_time-start_time, " seconds"
 
+! STEP 4: call the main solver routine. This solves the primal linear
+! system and computes the energy. If the forces are needed, it also
+! solves the adjoint linear system and computes the solute aspecific
+! contributions to the forces.
+if (ddx_data % params % force .eq. 1) then
+    allocate(force(3, ddx_data % params % nsph), stat=info)
+    if (info .ne. 0) then
+        write(6, *) "Allocation failed in ddx_driver"
+        stop 1
+    end if
+end if
 start_time = omp_get_wtime()
-call ddsolve(ddx_data, state, phi_cav, gradphi_cav, hessianphi_cav, psi, &
```

Comment:
Doesn't this imply that ddsolve is used nowhere? Maybe we should remove it completely?

Reply:
It is still used by the tests. I will remove it entirely then.

Reply:
Ok I see. To me the tests don't count as there we are sort of "free to call whatever function we want".